### DISCUSSION TITLE
xView - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#3040] [Docs]
# xView - Ultralytics YOLOv8 Docs

Discover the xView Dataset, a large-scale overhead imagery dataset for object detection tasks, featuring 1M instances, 60 classes, and high-resolution images.

https://docs.ultralytics.com/datasets/detect/xview/

**Comment by aHahii** [comment]
How to train the model to a good level? How to set the parameters?

**Reply by glenn-jocher** [reply]
@aHahii training a YOLOv8 model to a good level involves careful dataset preparation, parameter tuning, and possibly experimenting with different training strategies. Here are some general steps to follow:

1. **Prepare Your Dataset**: Ensure your dataset is well-labeled and representative of the problem you're trying to solve. Split it into training, validation, and test sets.

2. **Choose a Pretrained Model**: Starting with a pretrained model can significantly reduce training time and improve performance, especially if you have a limited dataset.

3. **Set Training Parameters**: Key parameters include batch size, learning rate, and the number of epochs. A larger batch size can provide more stable gradient estimates, but be mindful of your GPU memory limits. The learning rate controls how much the model weights are updated during training. A learning rate scheduler can be used to decrease the learning rate as training progresses. The number of epochs should be high enough for the model to converge but not so high that it overfits.

4. **Augmentation**: Use data augmentation to increase the diversity of your training data, which can help improve model robustness.

5. **Regularization**: Techniques like weight decay can help prevent overfitting.

6. **Monitor Training**: Use the validation set to monitor the model's performance. Look at metrics like precision, recall, and the mean Average Precision (mAP) to gauge how well your model is doing.

7. **Hyperparameter Tuning**: Experiment with different hyperparameters to find the best combination for your specific dataset and problem.

8. **Early Stopping**: Implement early stopping to terminate training when the validation loss stops improving.

For more detailed guidance on each of these steps, please refer to the training section in our documentation at https://docs.ultralytics.com/modes/train. Remember, achieving a well-performing model often requires experimentation and patience. Good luck! 🚀

**Comment by ttn0415** [comment]
Hi,

I have downloaded the xView dataset and the Yaml as specified and I am getting the following error:
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
[/Users/yolo/Desktop/xView/xView.ipynb](https://file+.vscode-resource.vscode-cdn.net/Users/yolo/Desktop/xView/xView.ipynb) Cell 6 line 4
      [1](vscode-notebook-cell:/Users/yolo/Desktop/xView/xView.ipynb#W5sZmlsZQ%3D%3D?line=0) #TensorBoard
      [2](vscode-notebook-cell:/Users/yolo/Desktop/xView/xView.ipynb#W5sZmlsZQ%3D%3D?line=1) #tensorboard --logdir runs/detect/train12
      [3](vscode-notebook-cell:/Users/yolo/Desktop/xView/xView.ipynb#W5sZmlsZQ%3D%3D?line=2) # Train the model
----> [4](vscode-notebook-cell:/Users/yolo/Desktop/xView/xView.ipynb#W5sZmlsZQ%3D%3D?line=3) results = model.train(data=f'{fpath}xView.yaml', epochs=50, imgsz=640,workers=24,batch=4,device='mps')

File [/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/ultralytics/engine/model.py:347](https://file+.vscode-resource.vscode-cdn.net/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/ultralytics/engine/model.py:347), in Model.train(self, trainer, **kwargs)
    345     self.model = self.trainer.model
    346 self.trainer.hub_session = self.session  # attach optional HUB session
--> 347 self.trainer.train()
    348 # Update model and cfg after training
    349 if RANK in (-1, 0):

File [/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/ultralytics/engine/trainer.py:195](https://file+.vscode-resource.vscode-cdn.net/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/ultralytics/engine/trainer.py:195), in BaseTrainer.train(self)
    192         ddp_cleanup(self, str(file))
    194 else:
--> 195     self._do_train(world_size)

File [/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/ultralytics/engine/trainer.py:390](https://file+.vscode-resource.vscode-cdn.net/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/ultralytics/engine/trainer.py:390), in BaseTrainer._do_train(self, world_size)
    387 final_epoch = (epoch + 1 == self.epochs) or self.stopper.possible_stop
    389 if self.args.val or final_epoch:
--> 390     self.metrics, self.fitness = self.validate()
    391 self.save_metrics(metrics={**self.label_loss_items(self.tloss), **self.metrics, **self.lr})
...
    145     self.metrics.process(*stats)
--> 146 self.nt_per_class = np.bincount(stats[-1].astype(int), minlength=self.nc)  # number of targets per class
    147 return self.metrics.results_dict

IndexError: list index out of range
Output is truncated. View as a [scrollable element](command:cellOutput.enableScrolling?93b0c437-6508-4432-8af4-db50902983f5) or open in a [text editor](command:workbench.action.openLargeOutput?93b0c437-6508-4432-8af4-db50902983f5). Adjust cell output [settings](command:workbench.action.openSettings?%5B%22%40tag%3AnotebookOutputLayout%22%5D)...

What is the issue?? It appears the error is somewhere in the ultralytics code, but I can't seem to find it.

**Reply by glenn-jocher** [reply]
@ttn0415 hello,

The error you're encountering, `IndexError: list index out of range`, typically indicates that the code is trying to access an index in a list that doesn't exist. This could be due to a mismatch between the expected data format and the actual data provided by the xView dataset.

To resolve this issue, please ensure the following:
1. The dataset is correctly downloaded and structured as expected by YOLOv8.
2. The `xView.yaml` file accurately reflects the dataset's path and structure.
3. The class indices in your annotations are within the correct range and match the number of classes defined in your `xView.yaml`.

If the problem persists, you may want to review the dataset and YAML file for any discrepancies. For further guidance on dataset preparation and troubleshooting, please refer to our documentation on data preparation and common issues.

**Comment by taesanbeck** [comment]
I have tried to use this xView dataset, but because it is not in YOLO Format it is really a pain to use. I can see the labels are originally in geojson format and some of the annotation coordinates are not needed which which is why the -1 is used in the YAML ( I think it was like id 75 and 83), So I tried to write my own python code to extract the annotations but had little success. When I went back to putting my ipynb file in the same directory as the unzipped data and just followed ultralytics instructions on this site I still got an error and the YAML apparently can't see the images here is my YAML:
# Ultralytics YOLO 🚀, AGPL-3.0 license
# DIUx xView 2018 Challenge https://challenge.xviewdataset.org by U.S. National Geospatial-Intelligence Agency (NGA)
# --------  DOWNLOAD DATA MANUALLY and jar xf val_images.zip to 'datasets/xView' before running train command!  --------
# Example usage: yolo train data=xView.yaml
# parent
# ├── ultralytics
# └── datasets
#     └── xView  ← downloads here (20.7 GB)


# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
path: /content/drive/MyDrive/datasets/xView/  # dataset root dir
train: images/autosplit_train.txt  # train images (relative to 'path') 90% of 847 train images
val: images/autosplit_val.txt # train images (relative to 'path') 10% of 847 train images



# Classes
names:
  0: Fixed-wing Aircraft
  1: Small Aircraft
  2: Cargo Plane
  3: Helicopter
  4: Passenger Vehicle
  5: Small Car
  6: Bus
  7: Pickup Truck
  8: Utility Truck
  9: Truck
  10: Cargo Truck
  11: Truck w/Box
  12: Truck Tractor
  13: Trailer
  14: Truck w/Flatbed
  15: Truck w/Liquid
  16: Crane Truck
  17: Railway Vehicle
  18: Passenger Car
  19: Cargo Car
  20: Flat Car
  21: Tank car
  22: Locomotive
  23: Maritime Vessel
  24: Motorboat
  25: Sailboat
  26: Tugboat
  27: Barge
  28: Fishing Vessel
  29: Ferry
  30: Yacht
  31: Container Ship
  32: Oil Tanker
  33: Engineering Vehicle
  34: Tower crane
  35: Container Crane
  36: Reach Stacker
  37: Straddle Carrier
  38: Mobile Crane
  39: Dump Truck
  40: Haul Truck
  41: Scraper/Tractor
  42: Front loader/Bulldozer
  43: Excavator
  44: Cement Mixer
  45: Ground Grader
  46: Hut/Tent
  47: Shed
  48: Building
  49: Aircraft Hangar
  50: Damaged Building
  51: Facility
  52: Construction Site
  53: Vehicle Lot
  54: Helipad
  55: Storage Tank
  56: Shipping container lot
  57: Shipping Container
  58: Pylon
  59: Tower



Here is my CODE(in Colab Pro with A100 GPU):
# Go to Edit > Notebook Settings and pick GPU for Hardware accelerator. Click save
%pip install -U ultralytics

from google.colab import drive

drive.mount('/content/drive', force_remount=True)

gdrive_dir = '/content/drive/MyDrive/datasets/xView'

from ultralytics import YOLO

# Load a model
model = YOLO('yolov8n.pt')  # load a pretrained model (recommended for training)

#TensorBoard
#tensorboard --logdir runs/detect/train12
# Train the model
results = model.train(data=f'{gdrive_dir}/xView.yaml', epochs=50, imgsz=640,workers=96,batch=8)

and then the ERROR:
Ultralytics YOLOv8.0.200 🚀 Python-3.10.12 torch-2.0.1+cu118 CUDA:0 (NVIDIA A100-SXM4-40GB, 40514MiB)
engine/trainer: task=detect, mode=train, model=yolov8n.pt, data=/content/drive/MyDrive/datasets/xView/xView.yaml, epochs=50, patience=50, batch=8, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=96, project=None, name=train, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, show=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, vid_stride=1, stream_buffer=False, line_width=None, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, boxes=True, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, cfg=None, tracker=botsort.yaml, save_dir=runs/detect/train
Downloading https://ultralytics.com/assets/Arial.ttf to '/root/.config/Ultralytics/Arial.ttf'...
100%|██████████| 755k/755k [00:00<00:00, 68.3MB/s]
Overriding model.yaml nc=80 with nc=60

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      7360  ultralytics.nn.modules.block.C2f             [32, 32, 1, True]             
  3                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                
  4                  -1  2     49664  ultralytics.nn.modules.block.C2f             [64, 64, 2, True]             
  5                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]               
  6                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]           
  7                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]              
  8                  -1  1    460288  ultralytics.nn.modules.block.C2f             [256, 256, 1, True]           
  9                  -1  1    164608  ultralytics.nn.modules.block.SPPF            [256, 256, 5]                 
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 11             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 12                  -1  1    148224  ultralytics.nn.modules.block.C2f             [384, 128, 1]                 
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 15                  -1  1     37248  ultralytics.nn.modules.block.C2f             [192, 64, 1]                  
 16                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
 17            [-1, 12]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 18                  -1  1    123648  ultralytics.nn.modules.block.C2f             [192, 128, 1]                 
 19                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
 20             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 21                  -1  1    493056  ultralytics.nn.modules.block.C2f             [384, 256, 1]                 
 22        [15, 18, 21]  1    763012  ultralytics.nn.modules.head.Detect           [60, [64, 128, 256]]          
Model summary: 225 layers, 3022548 parameters, 3022532 gradients, 8.3 GFLOPs

Transferred 319/355 items from pretrained weights
TensorBoard: Start with 'tensorboard --logdir runs/detect/train', view at http://localhost:6006/
Freezing layer 'model.22.dfl.conv.weight'
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks passed ✅
train: Scanning /content/drive/MyDrive/datasets/xView/labels/train... 0 images, 760 backgrounds, 0 corrupt: 100%|██████████| 760/760 [02:00<00:00,  6.32it/s]
train: WARNING ⚠️ No labels found in /content/drive/MyDrive/datasets/xView/labels/train.cache. See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.
train: New cache created: /content/drive/MyDrive/datasets/xView/labels/train.cache
WARNING ⚠️ No labels found in /content/drive/MyDrive/datasets/xView/labels/train.cache, training may not work correctly. See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01), CLAHE(p=0.01, clip_limit=(1, 4.0), tile_grid_size=(8, 8))
val: Scanning /content/drive/MyDrive/datasets/xView/labels/train... 0 images, 86 backgrounds, 0 corrupt: 100%|██████████| 86/86 [00:12<00:00,  6.68it/s]
val: WARNING ⚠️ No labels found in /content/drive/MyDrive/datasets/xView/labels/train.cache. See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.
val: New cache created: /content/drive/MyDrive/datasets/xView/labels/train.cache
WARNING ⚠️ No labels found in /content/drive/MyDrive/datasets/xView/labels/train.cache, training may not work correctly. See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.
Plotting labels to runs/detect/train/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.000156, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
zero-size array to reduction operation maximum which has no identity
Image sizes 640 train, 640 val
Using 8 dataloader workers
Logging results to runs/detect/train
Starting training for 50 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       1/50      1.16G          0       62.7          0          0        640: 100%|██████████| 95/95 [01:32<00:00,  1.03it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:01<00:00,  5.21it/s]
                   all         86          0          0          0          0          0
WARNING ⚠️ no labels found in detect set, can not compute metrics without labels

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       2/50      1.26G          0      59.59          0          0        640: 100%|██████████| 95/95 [00:10<00:00,  9.03it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:00<00:00, 40.22it/s]
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-4-728ec05913a2> in <cell line: 4>()
      2 #tensorboard --logdir runs/detect/train12
      3 # Train the model
----> 4 results = model.train(data=f'{gdrive_dir}/xView.yaml', epochs=50, imgsz=640,workers=96,batch=8)

6 frames
/usr/local/lib/python3.10/dist-packages/ultralytics/models/yolo/detect/val.py in get_stats(self)
    144         if len(stats) and stats[0].any():
    145             self.metrics.process(*stats)
--> 146         self.nt_per_class = np.bincount(stats[-1].astype(int), minlength=self.nc)  # number of targets per class
    147         return self.metrics.results_dict
    148 

IndexError: list index out of range

I am not sure why I am getting this error but I see the word 'self' so I know there is an error in some model Class structure that doesn't work with the way the YAML file is configured. I am trying to train YOLOV8 on Xview to obtatin the best.pt file to then retrain another YOLOv8 with xView Pretrained weights and my custom satellite imagery dataset my team is building for a class project at George Mason University and I would really appreciate any help you can provide.

**Reply by glenn-jocher** [reply]
@taesanbeck hello! It seems like you're encountering issues with the xView dataset and the YAML configuration for training YOLOv8. The error messages indicate that no labels were found in your dataset, which is likely the root cause of the problem.

Here are a few steps to troubleshoot and resolve the issue:

1. **Dataset Structure**: Ensure that your dataset is structured correctly with the images and labels in the expected directories. The labels should be in YOLO format and located in a `labels` folder parallel to the `images` folder.

2. **YAML Paths**: Verify that the paths specified in your YAML file are correct and accessible from your current working directory. The `path` should point to the root directory of the xView dataset, and the `train` and `val` should point to the text files containing the paths to your training and validation images.

3. **Label Conversion**: If your labels are in geojson format, you'll need to convert them to YOLO format. Make sure that this conversion is done correctly, and the resulting label files are placed in the correct directory.

4. **Cache Files**: The warning about no labels found in the cache file suggests that the dataset was not properly processed. You might need to clear any existing cache files and let the YOLOv8 create new ones during the training process.

5. **Dataset Check**: Use the `YOLO` class to load your dataset before training to check if it can correctly read the images and labels. This can help identify any issues with the dataset itself.

6. **Permissions**: If you're using Google Colab and mounting Google Drive, ensure that you have the correct permissions to access and modify the files in the dataset directory.

7. **YAML Content**: Double-check the content of your YAML file for any typos or formatting errors that might cause the dataset to be read incorrectly.

8. **Documentation**: Review the dataset formatting guidance in the Ultralytics Docs to ensure that your dataset meets the required format for training.

By following these steps, you should be able to identify where the issue lies and correct it. If you continue to experience problems, please provide a detailed description of the steps you've taken and any error messages you're receiving, and we'll do our best to assist you further. Good luck with your project at George Mason University! 🚀

**Reply by pollyyyyyyyyyyyyyyyyyyy** [reply]
Hi,

I am experiencing persistent issues while training YOLOv8 (YOLO11n model) with the xView dataset. Despite carefully following all the documented steps for dataset preparation and training, the results are consistently problematic. Here are the details:

Dataset Preparation
Resized Images:
I resized all images from the xView dataset to maintain a maximum size of 640 pixels on the largest dimension, ensuring compatibility with the imgsz=640 training parameter. The aspect ratio of the images was preserved during resizing.

Train/Validation/Test Split:
The xView training images and labels were divided as follows:

80% Train: For training the model.
10% Validation: For validation during training.
10% Test: Reserved for evaluation.
I ensured that each split had matching .txt label files for all corresponding images.

Labels Handling:

I used the official xView class mapping (classes 11-94 mapped to 0-59).
For images without annotations, I explicitly generated empty .txt files to represent unannotated images (backgrounds).
However, there are no true background images in the dataset—all images contain annotations except for a few explicitly empty .txt files added during preprocessing.
No Background Images:
After careful inspection:

No images without labels ("true backgrounds") were found in the xView dataset.
Even the potentially unannotated images (e.g., .tif files like ._100.tif) were checked, and they were either valid or excluded from training.
Training Details
Command Used:
python
model.train(data="xView.yaml", epochs=100, imgsz=640, batch=4)
Model: YOLO11n
Environment: Google Colab (A100 GPU)
Issues Observed
Metrics and Performance:

After multiple training attempts, the performance metrics remain extremely poor:
Precision: Close to 0 for most classes.
Recall: Close to 0, even for classes with a significant number of instances, such as "Small Car" and "Building."
mAP: Extremely low across the board.
Confusion Matrix:
The confusion matrix indicates that the model struggles to make meaningful predictions. It seems unable to learn effectively from the dataset, with predictions often biased toward specific classes or entirely absent for others.

No Background Handling:
Since there are no true background images, it is unclear whether the lack of explicit background examples is causing the model's poor generalization. Despite creating empty .txt files for unannotated images, the training results remain poor.

Steps Taken
Verified the dataset integrity:
Ensured all images have corresponding .txt labels.
Checked for corrupted files, finding none.
Resized the images to ensure compatibility with the imgsz=640 parameter.
Split the dataset into training, validation, and test sets with a clear 80/10/10 ratio.



Are there additional preprocessing steps or considerations for the xView dataset when training with YOLO11n?
Could the lack of true background images in the dataset be affecting performance?
Are there known compatibility issues between the xView dataset and YOLOv8?
What could explain the extremely poor performance metrics despite following the standard training pipeline?
Thank you in advance for your assistance. Any guidance or suggestions would be greatly appreciated!

**Reply by glenn-jocher** [reply]
Thank you for the detailed explanation of the steps you've taken. Based on your description, here are some suggestions:

1. Ensure that the `xView.yaml` file paths for `train`, `val`, and `test` are correct and relative to the `path` directory. Double-check that `autosplit_train.txt` and `autosplit_val.txt` correctly list image paths.
2. Verify that your labels are in YOLO format (class x_center y_center width height) and correctly map xView classes (11-94 to 0-59 per the provided mapping).
3. If no true background images exist, the presence of empty `.txt` files should suffice. However, ensure these files are correctly named and paired with the corresponding images.
4. Poor metrics might result from insufficient training data diversity or imbalanced classes. Consider augmenting the dataset or rebalancing the class distributions.
5. There are no known compatibility issues between xView and YOLOv8, but ensure you follow the conversion script provided [here](https://docs.ultralytics.com/datasets/detect/xview/) for label preparation.

If issues persist, please confirm your dataset structure and share minimal reproducible details for further assistance.
